src/include/homestore/chunk_selector.h

```diff
@@ -0,0 +1,30 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2023 eBay Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <homestore/vchunk.h>
+
+namespace homestore {
+class Chunk;
+class ChunkSelector {
+public:
+    ChunkSelector() = default;
+    virtual void add_chunk(VChunk chunk) = 0;
+    virtual void foreach_chunks(std::function< void(VChunk&) >&& cb) = 0;
+    virtual Chunk* select(blk_count_t nblks, const blk_alloc_hints& hints) = 0;
```

Comment:
Is there a reason to return a raw pointer here? Can we make it a unique ptr? From API signature, it means the caller of select now owns this chunk and when adding back the chunk it also can be a unique ptr, saying, hey, the caller is now passing ownership back to ChunkSelector.

Reply:
it could be shared_ptr but cannot be unique_ptr,  the selected chunk will not be removed from ChunkSelector as the `for_each` still expect to touch every chunk , regarless if they are in-use (remember the dicsussion to have for_each_active ).

Reply:
i just copy the current definition of chunk_seletcor, it is defined in the current.
actually, we can not return a unique_prt, since the since chunk_selector will also keep the pointer to it.

we can return  share_ptr if necessary